File: next/dweb-browser/src/Platforms/iOS/MicroModule/Browser/NativeUI/StatusBar/StatusBarNMM.cs
namespace DwebBrowser.MicroService.Browser.NativeUI.StatusBar;

public class StatusBarNMM : NativeMicroModule
{
    static readonly Debugger Console = new("StatusBarNMM");
    public StatusBarNMM() : base("status-bar.nativeui.browser.dweb")
    {
    }

    private Task<StatusBarController> _getControllerAsync(Mmid mmid) =>
        MainThread.InvokeOnMainThreadAsync(() => NativeUiController.FromMultiWebView(mmid).StatusBar);

    protected override async Task _bootstrapAsync(IBootstrapContext bootstrapContext)
    {
        HttpRouter.AddRoute(IpcMethod.Get, "/getState", async (_, ipc) =>
        {
            return await _getControllerAsync(ipc.Remote.Mmid);
        });

        HttpRouter.AddRoute(IpcMethod.Get, "/setState", async (request, ipc) =>
        {
            var controller = await _getControllerAsync(ipc.Remote.Mmid);
            request.QueryColor("color")?.Also(it =>
            {
                controller.ColorState.Set(it);
                //Console.Log("setState", "red: {0}, green: {1}, blue: {2}, alpha: {3}",
                //    it.red, it.green, it.blue, it.alpha);
                controller.Observer.Get();
            });
            request.QueryStyle("style")?.Also(it =>
            {
                controller.StyleState.Set(it);
                controller.Observer.Get();
            });
            request.QueryBool("overlay")?.Also(it =>
            {
                controller.OverlayState.Set(it);
                controller.Observer.Get();
            });
            request.QueryBool("visible")?.Also(it =>
            {
                controller.VisibleState.Set(it);
                controller.Observer.Get();
            });
            return null;
        });

        HttpRouter.AddRoute(IpcMethod.Get, "/startObserve", async (_, ipc) =>
        {
            return (await _getControllerAsync(ipc.Remote.Mmid)).StateObserver.StartObserve(ipc);
        });

        HttpRouter.AddRoute(IpcMethod.Get, "/stopObserve", async (_, ipc) =>
        {
            return (await _getControllerAsync(ipc.Remote.Mmid)).StateObserver.StopObserve(ipc);
        });
    }
}

File: android/app/src/main/java/info/bagen/dwebbrowser/microService/browser/nativeui/statusBar/StatusBarNMM.kt
package info.bagen.dwebbrowser.microService.browser.nativeui.statusBar

import info.bagen.dwebbrowser.microService.browser.nativeui.NativeUiController
import info.bagen.dwebbrowser.microService.browser.nativeui.helper.QueryHelper
import info.bagen.dwebbrowser.microService.browser.nativeui.helper.fromMultiWebView
import io.ktor.http.HttpMethod
import io.ktor.http.HttpStatusCode
import org.dweb_browser.helper.toJsonElement
import org.dweb_browser.microservice.core.BootstrapContext
import org.dweb_browser.microservice.core.NativeMicroModule
import org.dweb_browser.microservice.help.types.MICRO_MODULE_CATEGORY
import org.dweb_browser.microservice.help.types.MMID
import org.dweb_browser.microservice.http.PureResponse
import org.dweb_browser.microservice.http.PureStreamBody
import org.dweb_browser.microservice.http.bind
import org.dweb_browser.microservice.http.routes

class StatusBarNMM : NativeMicroModule("status-bar.nativeui.browser.dweb", "statusBar") {

  init {
    categories = listOf(MICRO_MODULE_CATEGORY.Service, MICRO_MODULE_CATEGORY.Render_Service)
  }

  private fun getController(mmid: MMID) = NativeUiController.fromMultiWebView(mmid).statusBar

  override suspend fun _bootstrap(bootstrapContext: BootstrapContext) {
    routes(
      /** 获取状态栏 */
      "/getState" bind HttpMethod.Get to defineJsonResponse {
        return@defineJsonResponse getController(ipc.remote.mmid).toJsonElement()
      },
      /** 设置状态栏 */
      "/setState" bind HttpMethod.Get to defineEmptyResponse {
        val controller = getController(ipc.remote.mmid)
        QueryHelper.color(request)?.also { controller.colorState.value = it }
        QueryHelper.style(request)?.also { controller.styleState.value = it }
        QueryHelper.overlay(request)?.also { controller.overlayState.value = it }
        QueryHelper.visible(request)?.also { controller.visibleState.value = it }
        return@defineEmptyResponse
      },
      /**
       * 开始数据订阅
       */
      "/observe" bind HttpMethod.Get to definePureResponse {
        val inputStream = getController(ipc.remote.mmid).observer.startObserve(ipc)
        return@definePureResponse PureResponse(
          HttpStatusCode.OK, body = PureStreamBody(inputStream)
        )
      },
    )
//      .cors()
  }

  override suspend fun _shutdown() {
    TODO("Not yet implemented")
  }
}
